connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/functions.scala

```diff
@@ -72,51 +69,129 @@ object functions {
   }
 
   /**
-   * Converts a binary column of Protobuf format into its corresponding catalyst value. The
-   * specified Protobuf class must match the data, otherwise the behavior is
-   * undefined: it may fail or return arbitrary result. The jar containing Java class should be
+   * Converts a binary column of Protobuf format into its corresponding catalyst value.
+   * `messageClassName` points to Protobuf Java class. The jar containing Java class should be
    * shaded. Specifically, `com.google.protobuf.*` should be shaded to
    * `org.sparkproject.spark-protobuf.protobuf.*`.
+   * https://github.com/rangadi/shaded-protobuf-classes is useful to create shaded jar from
+   * Protobuf files.
    *
    * @param data
    *   the binary column.
-   * @param shadedMessageClassName
-   *   The Protobuf class name. E.g. <code>org.spark.examples.protobuf.ExampleEvent</code>.
+   * @param messageClassName
+   *   The full name for Protobuf Java class. E.g. <code>com.example.protos.ExampleEvent</code>.
    *   The jar with these classes needs to be shaded as described above.
    * @since 3.4.0
    */
   @Experimental
-  def from_protobuf(data: Column, shadedMessageClassName: String): Column = {
-    new Column(ProtobufDataToCatalyst(data.expr, shadedMessageClassName))
+  def from_protobuf(data: Column, messageClassName: String): Column = {
+    new Column(ProtobufDataToCatalyst(data.expr, messageClassName))
   }
 
   /**
-   * Converts a column into binary of protobuf format.
+   * Converts a binary column of Protobuf format into its corresponding catalyst value.
+   * `messageClassName` points to Protobuf Java class. The jar containing Java class should be
+   * shaded. Specifically, `com.google.protobuf.*` should be shaded to
+   * `org.sparkproject.spark-protobuf.protobuf.*`.
+   * https://github.com/rangadi/shaded-protobuf-classes is useful to create shaded jar from
+   * Protobuf files.
+   *
+   * @param data
+   *   the binary column.
+   * @param messageClassName
+   *   The full name for Protobuf Java class. E.g. <code>com.example.protos.ExampleEvent</code>.
+   *   The jar with these classes needs to be shaded as described above.
+   * @param options
+   * @since 3.4.0
+   */
+  @Experimental
```

Comment:
Just curious, will spark in future considering adding a dataSource for Protobuf files? Supporting read/write from and to Protobuf.

Reply:
@ericsun95 since the protobuf encoding format is binary, you can use the spark binary file data source and then use from_protobuf function to convert it into a spark dataframe. I don't see the need for a separate protobuf data source. 

`
spark.read.format("binaryFile").option("pathGlobFilter", "*.proto").load("/path/to/data")
`

having a separate data source for protobuf can achieve performance, type safety, and schema evolution by storing the protobuf schema along with the protobuf files on disk.

@rangadi

Reply:
That is actually good to now, I will try it. Are you saying it is similarly if I write it to protobuf as well? No need a extra dataSource for it?

Reply:
I tried, I found it didn't work? I can get the binary but when I add the `.option("pathGlobFilter", "*path_to/*.proto")` it actually return empty dataframe.

 Also, the real value is here, for example, a large protobuf file can be XX GBs. If we don't implement a datasource which split the file into partitions. This is really hard to scale.

Reply:
@ericsun95 Protobuf files can be in either .pb or binary format. .proto files contain the schema for the protobuf data, while binary files contain the actual data. If your protobuf files are in binary format, you can use the Spark binaryFile data source to read them.

Reply:
There is no "protobuf file format" (unlike Avro, which has its own format).
One common way I have seen are "length delimited files" where the record length is written followed by the record.
We can use "binaryFile" to handle such files. But there some disadvantages: 
  - It loads the whole file in to memory 
  - User needs to parse the length and the extract the bytes.

It will be good to add a new format "delimitedFile" that handles such file and return one binary record at a time.

Reply:
Yeah, that would also be good. One example would be openstreetmap data (.osm.pbf > 60GB) that makes loading the whole file in memory is infeasible. I can see that [scalapb](https://scalapb.github.io/docs/sparksql/) has some integration with spark while it also loads the whole file into memory. Unless we create our own datasources with customized logic, using spark with protobuf is not as easy as other file formats right now.

Reply:
What is the format of OpenStreetMap data? 
This would be a good contribution to open source spark. The implementation will be similar to "binaryFile" format implementation.

Reply:
https://wiki.openstreetmap.org/wiki/PBF_Format. You can see the protoc here  https://github.com/openstreetmap/OSM-binary/tree/master/osmpbf

Reply:
@rangadi  After doing some initial reading on the .pbf file format, I understand that it is a binary file format that uses zlib compression to compress protobuf format data. This means that in order to read a .pbf file, we need to first decompress the data. 

- If we want to break a big .pdf file into N chunks, we can use the .pdf BlobHeader to determine the size of the file and then split it into N equal-sized chunks. However, this will not work if the whole .pdf file is a single large protobuf value. In this case, we will need to use a different approach to break the file into chunks.

- One approach is to use the protobuf schema to break the file into chunks. The protobuf schema is a file that describes the structure of the protobuf data. We can use the protobuf schema to identify the different fields in the protobuf data and then break the file into chunks based on these fields.

- Another approach is to use a streaming parser to break the file into chunks.

- Once we have broken the file into chunks, we can distribute the chunks to different Spark executors. We can then parse the chunks into Spark DataFrames.

Reply:
@rangadi @ericsun95 I was able to get it to work with the binaryFile format by making a few minor changes to the from_protobuf module.

```
val df spark.read.format("binaryFile").load("git/OSM-binary/resources/")

val protoDf = df.withColumn( "Headerblock", 
org.apache.spark.sql.protobuf.functions.from_protobuf(col("content"), "HeaderBlock", descFile))
.withColumn("DenseInfo",
org.apache.spark.sql.protobuf.functions.from_protobuf( col("content"), "DenseInfo", descFile)).drop("content").drop("length").drop("modificationTime").drop("path")

protoDf.show(truncate = false)
```

and The ProtobufDataToCatalyst class also needs to be tweaked a bit at the [line](https://github.com/apache/spark/blob/master/connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDataToCatalyst.scala#L103)

```
if (buffer.get() == MAGIC_BYTE) {
  val datinput: DataInputStream = new DataInputStream(new ByteArrayInputStream(buffer.array()))
  val headersize = datinput.readInt()
  if (headersize > MAX_HEADER_SIZE) {
    throw new FileFormatException("Unexpectedly long header " + MAX_HEADER_SIZE + " bytes. Possibly corrupt file.");
  }
  binary = new Array[Byte](headersize)
  datinput.readFully(binary)

  val blogHeader = BlobHeader.parseFrom(binary)

  val out : FileBlock  = FileBlock.newInstance(blogHeader.getType, null, blogHeader.getIndexdata)
  val binaryData: Array[Byte] = new Array[Byte](blogHeader.getDatasize)
  datinput.readFully(binaryData)

  val blob = Blob.parseFrom(binaryData)

  if (blob.hasRaw()) {
      binary = blob.getRaw().toByteArray;
  } else if (blob.hasZlibData()) {
    val buf2: Array[Byte]  = new  Array[Byte](blob.getRawSize())
    val decompresser: Inflater  = new Inflater();
    decompresser.setInput(blob.getZlibData().toByteArray());
    try {
      decompresser.inflate(buf2);
    } catch {
      case e: DataFormatException =>
        throw new UncheckedIOException(new FileFormatException(e));
    }
    assert(decompresser.finished());
    decompresser.end();
    binary = ByteString.copyFrom(buf2).toByteArray
  }
```

@rangadi would it be useful to have a spark source for .pbf files? I would be happy to work on it.

Reply:
@SandishKumarHN I am not sure about 'pbf' file. That seems quite specific to OpenStreetMaps. We can ask on spark-dev/user list. 

If you are interested, we have a few improvements we can do for for protobuf functions. let me know.

Reply:
> @rangadi After doing some initial reading on the .pbf file format, I understand that it is a binary file format that uses zlib compression to compress protobuf format data. This means that in order to read a .pbf file, we need to first decompress the data.
> 
> * If we want to break a big .pdf file into N chunks, we can use the .pdf BlobHeader to determine the size of the file and then split it into N equal-sized chunks. However, this will not work if the whole .pdf file is a single large protobuf value. In this case, we will need to use a different approach to break the file into chunks.
> * One approach is to use the protobuf schema to break the file into chunks. We can use the protobuf schema to identify the different fields in the protobuf data and then break the file into chunks based on these fields.
> * Another approach is to use a streaming parser to break the file into chunks.
> * Once we have broken the file into chunks, we can distribute the chunks to different Spark executors. We can then parse the chunks into Spark DataFrames.

Yeah you are right. Previously we have to write customized parser with spark data source interface to make it distributable https://github.com/ericsun95/spark-osm-pbf/tree/master/src/main/scala/com/ericsun95/spark/osm/pbf/parser. However, I think it would be better if we can have generic way for such kind of use case. That would benefit all kinds of user of protobuf. We can either enhance the binary file format (more compatible, write support, dataSourceV2 support etc) or have a standalone datasource. I think will have lots of buyings from the community.

Reply:
> @SandishKumarHN I am not sure about 'pbf' file. That seems quite specific to OpenStreetMaps. We can ask on spark-dev/user list.
> 
> If you are interested, we have a few improvements we can do for for protobuf functions. let me know.
please assign me the JIRA's.